src/3d/qgs3dexportobject.cpp

```diff
@@ -0,0 +1,205 @@
+  Qgs3DExportObject.cpp
+  --------------------------------------
+  Date                 : June 2020
+  Copyright            : (C) 2020 by Belgacem Nedjima
+  Email                : gb underscore nedjima at esi dot dz
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dexportobject.h"
+
+#include <QVector3D>
+#include <QDebug>
+#include <QDir>
+#include <QImage>
+
+#include <Qt3DRender/QAttribute>
+#include <Qt3DRender/QBuffer>
+#include <Qt3DRender/QBufferDataGenerator>
+#include <Qt3DRender/QBufferDataGeneratorPtr>
+
+#include "qgsphongmaterialsettings.h"
+
+
+template<typename T>
+void insertIndexData( QVector<uint> &vertexIndex, const QVector<T> &faceIndex )
+{
+  for ( int i = 0; i < faceIndex.size(); i += 3 )
+  {
+    if ( i + 2 >= faceIndex.size() ) continue;
+    // skip invalid triangles
+    if ( faceIndex[i] == faceIndex[i + 1] || faceIndex[i + 1] == faceIndex[i + 2] || faceIndex[i] == faceIndex[i + 2] )
+      continue;
+    for ( int j = 0; j < 3; ++j )
+      vertexIndex << faceIndex[i + j] + 1;
+  }
+}
+
+Qgs3DExportObject::Qgs3DExportObject( const QString &name, const QString &parentName, QObject *parent )
+  : QObject( parent )
+  , mName( name )
+  , mParentName( parentName )
+  , mSmoothEdges( false )
+{
+}
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, float scale, const QVector3D translation )
+{
+  for ( int i = 0; i < positionsBuffer.size(); i += 3 )
+  {
+    for ( int j = 0; j < 3; ++j )
+    {
+      mVertexPosition << positionsBuffer[i + j] * scale + translation[j];
+    }
+  }
+}
+
+void Qgs3DExportObject::setupFaces( const QVector<uint> &facesIndexes )
+{
+  insertIndexData<uint>( mIndexes, facesIndexes );
+}
+
+void Qgs3DExportObject::setupLine( const QVector<uint> &lineIndexes )
+{
+  Q_UNUSED( lineIndexes );
+  for ( int i = 0; i < mVertexPosition.size(); i += 3 ) mIndexes << i / 3 + 1;
+}
+
+void Qgs3DExportObject::setupNormalCoordinates( const QVector<float> &normalsBuffer )
+{
+  mNormals << normalsBuffer;
+}
+
+void Qgs3DExportObject::setupTextureCoordinates( const QVector<float> &texturesBuffer )
+{
+  mTexturesUV << texturesBuffer;
+}
+
+void Qgs3DExportObject::setupMaterial( QgsAbstractMaterialSettings *material )
+{
+  QMap<QString, QString> parameters = material->toExportParameters();
+  for ( QString &key : parameters.keys() ) setMaterialParameter( key, parameters[key] );
```

Comment:
```suggestion
  const QMap<QString, QString> parameters = material->toExportParameters();
  for ( auto it = parameters.begin(); it != parameters.end(); ++it )
  {
     setMaterialParameter( it.key(), it.value() );
  }
```

(this is considerable more efficient -- it avoids the initial loop over the container to create the keys list, and then the n x map lookups for each key in the map)